fdbserver/storageserver.actor.cpp

```diff
@@ -1611,6 +1612,67 @@ void merge(Arena& arena,
 	}
 }
 
+ACTOR Future<Optional<Value>> quickGetValue(StorageServer* data, StringRef key, Version version) {
+	if (data->shards[key]->isReadable()) {
+		try {
+			// TODO: Use a lower level API may be better? Or tweak priorities?
+			GetValueRequest req(Span().context, key, version, Optional<TagSet>(), Optional<UID>());
+			data->actors.add(data->readGuard(req, getValueQ));
+			GetValueReply reply = wait(req.reply.getFuture());
+			return reply.value;
+		} catch (Error& e) {
+			std::cout << "quickGetValue fallback because of exception " << e.name() << std::endl;
+		}
+	} else {
+		std::cout << "quickGetValue fallback because not in the server" << std::endl;
+	}
+
+	state Transaction tr(data->cx);
+	tr.setVersion(version);
+	// TODO: is DefaultPromiseEndpoint the best priority for this?
+	tr.info.taskID = TaskPriority::DefaultPromiseEndpoint;
+	Future<Optional<Value>> valueFuture = tr.get(key, Snapshot::True);
+	// TODO: async in case it needs to read from other servers.
+	state Optional<Value> valueOption = wait(valueFuture);
+	return valueOption;
+};
+
+ACTOR Future<GetKeyValuesAndHopReply> hop(StorageServer* data, GetKeyValuesReply input, HopInfo hopInfo) {
+	state GetKeyValuesAndHopReply result;
+	result.version = input.version;
+	result.more = input.more;
+	result.cached = input.cached;
+	result.arena.dependsOn(input.arena);
+
+	result.data.reserve(result.arena, input.data.size());
+	state KeyValueRef* it = input.data.begin();
+	for (; it != input.data.end(); it++) {
+		state StringRef key = it->key;
+		Tuple keyTuple = Tuple::unpack(key);
+		size_t suffixStart = keyTuple.size() - hopInfo.suffixLen;
+		if (suffixStart < 0) {
+			throw hop_bad_suffix();
+		}
+		Tuple suffix = keyTuple.subTuple(suffixStart);
+
+		Standalone<StringRef> hopKey = Tuple().append(hopInfo.hopPrefix).append(suffix).getDataAsStandalone();
+		// Make sure the hopKey is always available, so that it's good even we want to get key asynchronously.
+		result.arena.dependsOn(hopKey.arena());
+
+		std::cout << "quickGetValue start key: " << hopKey.toString() << std::endl;
+		Optional<Value> valueOption = wait(quickGetValue(data, hopKey, input.version));
```

Comment:
If we can make sure the input.data indexed data is still on the same SS, quickGetValue() will not need the transaction inside it. right?

Reply:
Yes, see https://github.com/apple/foundationdb/pull/5609/files#diff-c4dbfae809f757f2061099ee5021ce2cbbf0f4dbc2f0f18480a4856d63d033ceR1619-R1622. Most of the queries should hit this **and return**, without transactions.